ops/testing.py

```diff
@@ -1314,25 +1317,292 @@ def get_services(self, names: typing.List[str] = None) -> typing.List[pebble.Ser
+    (e.g. \, \foo\bar, C:\foo\bar) are not supported.
+    """
+
+    def __init__(self):
+        self.root = _Directory(pathlib.PurePosixPath('/'))
+
+    def create_dir(self, path: str, make_parents: bool = False, **kwargs) -> '_Directory':
+        if not path.startswith('/'):
+            raise NonAbsolutePathError(path)
+        current_dir = self.root
+        tokens = pathlib.PurePosixPath(path).parts[1:]
+        for token in tokens[:-1]:
+            if token in current_dir:
+                current_dir = current_dir[token]
+            else:
+                if make_parents:
+                    # NOTE: other parameters (e.g. ownership, permissions) only get applied to the
+                    # final directory.
+                    # (At the time of writing, Pebble defaults to 0o755 permissions and root:root
+                    # ownership.)
+                    current_dir = current_dir.create_dir(token)
+                else:
+                    raise FileNotFoundError(str(current_dir.path / token))
+            if isinstance(current_dir, _File):
+                raise NotADirectoryError(str(current_dir.path))
+
+        # Current backend will always raise an error if the final directory component
+        # already exists.
+        token = tokens[-1]
+        if token not in current_dir:
+            current_dir = current_dir.create_dir(token, **kwargs)
+        else:
+            raise FileExistsError(str(current_dir.path / token))
+        return current_dir
+
+    def create_file(
+            self,
+            path: str,
+            data: typing.Union[bytes, str, typing.BinaryIO, typing.TextIO],
+            encoding: typing.Optional[str] = 'utf-8',
+            make_dirs: bool = False,
+            **kwargs
+    ) -> '_File':
+        if not path.startswith('/'):
+            raise NonAbsolutePathError(path)
+        path_obj = pathlib.PurePosixPath(path)
+        try:
+            dir_ = self.get_path(path_obj.parent)
+        except FileNotFoundError:
+            if make_dirs:
+                dir_ = self.create_dir(str(path_obj.parent))
+                # NOTE: other parameters (e.g. ownership, permissions) only get applied to the
+                # final directory.
+                # (At the time of writing, Pebble defaults to the specified permissions and
+                # root:root ownership, which is inconsistent with the push function's
+                # behavior for parent directories.)
+            else:
+                raise
+        if not isinstance(dir_, _Directory):
+            raise pebble.PathError(
+                'generic-file-error', 'parent is not a directory: {}'.format(str(dir_)))
+        return dir_.create_file(path_obj.name, data, encoding=encoding, **kwargs)
+
+    def list_dir(self, path) -> typing.List['_File']:
+        current_dir = self.root
+        tokens = pathlib.PurePosixPath(path).parts[1:]
+        for token in tokens:
```

Comment:
We should have a note here about checking permissions. I don't think it is a critical need today, but if I did want to write my charm and have it deal with permission errors, we would need that functionality.
But 'consider this in the future' is probably sufficient for now, as it is just testing infrastructure, and we don't actually have a request for this functionality yet.